### PR TITLE
Support the merge of decoder without/with past for encoder-decoder models in the ONNX export

### PR DESCRIPTION
**Pull request by fxmarty**
As per title, will need https://github.com/huggingface/optimum/pull/924 to be merged first.

Next PR: support this in ORTModel as well.

### CODE REVIEW

File: optimum/exporters/onnx/base.py [hunk 1]
```diff
@@ -614,6 +616,26 @@ def flatten_output_collection_property(self, name: str, field: Iterable[Any]) ->
...
 
+    def generate_dummy_inputs_for_validation(self, reference_model_inputs: Dict[str, Any]) -> Dict[str, Any]:
+        if self._behavior is ConfigBehavior.DECODER:
+            reference_model_inputs["input_ids"] = reference_model_inputs.pop("decoder_input_ids")
+            reference_model_inputs["encoder_hidden_states"] = reference_model_inputs.pop("encoder_outputs")[0]
+
+        if self.is_merged is True and self.use_cache_branch is True:
+            reference_model_inputs["use_cache_branch"] = DummyInputGenerator.constant_tensor(shape=[1], value=True)
+        elif self.is_merged is True and self.use_cache_branch is False:
+            reference_model_inputs["use_cache_branch"] = DummyInputGenerator.constant_tensor(shape=[1], value=False)
```
Comment (michaelbenayoun) [inline optimum/exporters/onnx/base.py:627]:
```suggestion
        if self.is_merged:
            reference_model_inputs["use_cache_branch"] = DummyInputGenerator.constant_tensor(shape=[1], value=self.use_cache_branch)
```

File: optimum/exporters/onnx/base.py [hunk 2]
```diff
@@ -731,7 +755,11 @@ def add_past_key_values(self, inputs_or_outputs: Dict[str, Dict[int, str]], dire
             inputs_or_outputs[f"{name}.{i}.decoder.key"] = {0: "batch_size", 2: decoder_sequence_name}
             inputs_or_outputs[f"{name}.{i}.decoder.value"] = {0: "batch_size", 2: decoder_sequence_name}
 
-            if direction == "inputs" or (self._behavior is ConfigBehavior.DECODER and self.use_past is False):
+            if (
+                direction == "inputs"
+                or (self._behavior is ConfigBehavior.DECODER and self.use_past is False)
+                or self.is_merged is True
```
Comment (michaelbenayoun) [inline optimum/exporters/onnx/base.py:761]: Small nit, but I would re-order those tests, for instance checking `self.is_merged` is True is much faster than `direction == "inputs"`.

File: optimum/exporters/onnx/base.py [hunk 3]
```diff
@@ -699,7 +723,7 @@ def outputs(self) -> Dict[str, Dict[int, str]]:
             new_axes_names = {}
             for axis_idx, axis_name in axes_names.items():
                 if "sequence" in axis_name:
-                    if not self.use_past_in_inputs:
+                    if self.use_past_in_inputs is False or self.is_merged is True:
```
Comment (michaelbenayoun) [inline optimum/exporters/onnx/base.py:726]: ...

**Review by michaelbenayoun — APPROVED** [review]
LGTM

File: optimum/exporters/onnx/base.py (hunk 2 above)
Comment (fxmarty) [reply optimum/exporters/onnx/base.py:757]: thanks!

File: optimum/exporters/onnx/base.py (hunk 1 above)
Comment (fxmarty) [reply optimum/exporters/onnx/base.py:623]: awesome

**Pull request merged**